src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp

```diff
@@ -286,90 +464,223 @@ class npc_stalagg : public CreatureScript
+                                me->SetInCombatWith(feugen->EnsureVictim());
+                            }
+                        powerSurgeTimer = urand(20, 25) * IN_MILLISECONDS;
+                        break;
+                    case ACTION_TRANSITION:
+                        if (me->IsAlive()) me->Kill(me); // sanity check just in case of weird race conditions
+
+                        if (myCoil() && myCoilGO())
+                        {
+                            myCoil()->CastStop();
+                            myCoilGO()->SetGoState(GO_STATE_ACTIVE);
+                            myCoil()->Talk(EMOTE_TESLA_OVERLOAD, CHAT_MSG_RAID_BOSS_EMOTE, 50000.0f, NULL);
+                        }
+                        break;
+                    case ACTION_TRANSITION_3:
+                        if (myCoilGO())
+                            myCoilGO()->SetGoState(GO_STATE_READY);
+                        break;
+                }
+            }
 
-        void UpdateAI(uint32 uiDiff) override
-        {
-            if (!UpdateVictim())
-                return;
+            void KilledUnit(Unit* /*victim*/) override
+            {
+                if (!(rand32() % 5))
+                    Talk(SAY_STALAGG_SLAY);
+            }
 
-            if (magneticPullTimer <= uiDiff)
+            void EnterCombat(Unit* who) override
             {
-                if (Creature* pFeugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
-                {
-                    Unit* pStalaggVictim = me->GetVictim();
-                    Unit* pFeugenVictim = pFeugen->GetVictim();
+                Talk(SAY_STALAGG_AGGRO);
+                if(!me->HasAura(SPELL_STALAGG_TESLA))
+                    DoCast(SPELL_STALAGG_TESLA);
 
-                    if (pFeugenVictim && pStalaggVictim)
+                if (Creature* thaddius = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_THADDIUS)))
+                    if (thaddius->AI())
+                        thaddius->AI()->DoAction(ACTION_STALAGG_AGGRO);
+
+                if(Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                    if (!feugen->IsInCombat())
                     {
-                        // magnetic pull is not working. So just jump.
+                        feugen->AddThreat(who, 0.0f);
+                        feugen->SetInCombatWith(who);
+                    }
+            }
 
-                        // reset aggro to be sure that feugen will not follow the jump
-                        pFeugen->getThreatManager().modifyThreatPercent(pFeugenVictim, -100);
-                        pFeugenVictim->JumpTo(me, 0.3f);
+            void DamageTaken(Unit* who, uint32& damage) override
+            {
+                if (damage < me->GetHealth()) return;
+                if (myCoil())
+                    myCoil()->CastStop();
+                if (who)
```

Comment:
*who* can never be null in this hook (at least there are no ways I can think of it being null)